### DISCUSSION TITLE
How to work with multiple container registries?

### DISCUSSION THREAD

**Discussion by Brightside56** [akuity/kargo#1981] [Q&A] [answered]
Hello

I have a case where there are multiple container registries used for same container image on different envs/stages. What means I can have prod-1.domain.com/app:0.1.0 and prod-2.domain.com/app:0.1.0 for different stages, but from what I see Kargo warehouses are designed to track single image (to promote it across stages). Existing promotion mechanisms imply a complete replacement of the image field in the manifest (without possibility to promote just an image tag, what could be a workaround).

Even if an organization logically uses one registry as build target for all container images for all environments, then when using cross-regional replication, registries such as AWS ECR will still produce different endpoints for different regions, and Kargo will treat them as different registries

How are such issues supposed to be solved with Kargo?

**Comment by krancour** [comment]
You are using Helm, iirc? The option to update a tag only exists. I'm out of picket atm, but the reference docs for the CRDs should say something about this.

**Reply by Brightside56** [reply]
This case (with umbrella charts and many artifacts) is on hold for now (because I believe Kargo is not ready for prime time here, but it may change soon)

Right now I'm making a PoC with pipeline per microservice (for a greenfield project which matches better, where freightline contains just git commit and one image built to ECR and replicated to different regions/accounts). Helm is used just for templating, render is used as a promotion mechanism. And I'm afraid there are no OOTB options, because kargo-render is not Helm values aware and doesn't have a possibility to distinguish image repositories where it could replace a tag =/

**Comment by krancour** [comment, marked as answer]
@Brightside56 Kargo Render passes images to the Argo CD Repo Server (which it uses as a library) i the form `url=tag`. Somewhere under the covers, the Argo CD Repo Server executes a `kustomize edit set image url=tag`. For Kargo's built-in Kustomize support, Kargo executes `kustomize edit set image url=tag` more directly...

I see your issue is that the image URL varies from Stage to Stage.

I believe your case can be addressed by adding more options to both Kargo Render and to Kargo's built-in Kustomize support such that the resulting command resembles this: `kustomize edit set image placeholder-url=stage-specific-url:tag` where `placeholder-url` and `stage-specific-url` are both things you specific in your promotion mechanisms on a Stage-by-Stage basis, which will leave the tag as the only part of the Freight that is applied from one Stage to the next.

I am happy to open an issue for this, but cannot guarantee it to be a priority.

**Reply by krancour** [reply]
I opened https://github.com/akuity/kargo/issues/2001
